lib/utils/gasEstimation.js

```diff
@@ -17,282 +15,178 @@ module.exports = (vm, runArgs, err, callback) => {
-      const oneSixtyFloorth = neededBefore6364 - sumOfCurrentSysOp;
-
-      systemOpSum = neededBefore6364 + callingFee;
-      const next = steps.findIndex(lowestIndex) + 1;
-
-      const prevParentBegin = parentBegin;
-      const prevParentEnd = parentEnd;
-
-      ({ parentBegin, parentEnd } = findParentScope(steps.systemOps[next][0]));
-      lowestIndex = parentEnd - 1;
-
-      // Sum to the top of the range above sysOp
-      systemOpSum += gas(steps.ops[parentBegin + 1]) - gas(steps.ops[prevParentBegin]);
-      // systemOpSum += sumRange(parentBegin - 1, prevParentBegin);
-
-      // Sum of the remaining steps after the current (outer) system operation
-      const costOfParentReturn =
-        gas(steps.ops[parentEnd - 1]) -
-        gas(steps.ops[parentEnd]) +
-        gas(steps.ops[parentBegin]) -
-        steps.ops[parentBegin].opcode.fee -
-        gas(steps.ops[parentBegin + 1]);
-      // recursively sum the remaining search space and return this total minus the 1/64th we already accounted for
-      const remaining = getTotal(prevParentEnd, lowestIndex) + costOfParentReturn;
-      // if ((stipend && stipend < remaining) || oneSixtyFloorth < remaining) {
-      let tail = oneSixtyFloorth;
-      if (callingSysOp.opcode.name === "CALL" && isStipend(steps.ops[rootBegin + 1])) {
-        tail = 2300;
-      }
-      if (tail < remaining) {
-        systemOpSum = 0;
+  const getTotal = () => {
+    const sysops = steps.systemOps;
+    const ops = steps.ops;
+    const opIndex = (cursor) => sysops[cursor].index;
+    const stack = [];
+    let cursor = 0;
+    let context = Context(0);
+    while (cursor < sysops.length) {
+      const currentIndex = opIndex(cursor);
+      const current = ops[currentIndex];
+      const name = current.opcode.name;
+      if (isCall(name)) {
+        if (steps.isSVT(currentIndex)) {
+          context.setStop(currentIndex + 1);
+          context.addRange(true);
+          context.setStart(currentIndex + 1);
+          context.addSixtyFloorth(2300);
+        } else {
+          context.setStop(currentIndex);
+          context.addRange(current.opcode.fee);
+          stack.push(context);
+          context = Context(currentIndex, current.opcode.fee); // setup next context
+        }
+      } else if (isCreate(name)) {
+        context.setStop(currentIndex);
+        context.addRange(current.opcode.fee);
+        stack.push(context);
+        context = Context(currentIndex, current.opcode.fee); // setup next context
+      } else if (isTerminator(name)) {
+        // only on the last operation
+        context.setStop(currentIndex + 1 < steps.ops.length ? currentIndex + 1 : currentIndex);
+        context.addRange();
+        const ctx = stack.pop();
+        if (ctx) {
+          ctx.transfer(context);
+          context = ctx;
+          context.setStart(currentIndex + 1);
+        } else {
+          let gas = context.getCost();
+          return gas.cost + gas.sixtyFloorths;
```

Comment:
I find it strange that we return within the while loop, but never outside the while loop.

We also seem to check if our `currentIndex` is the index of the last op, which _should_ be what the while loop's condition does (though I guess that is checking only sys ops).

I have some ideas we can explore to better handle the "when are we done case" that involves changing the way start/end are set and the way `addRange` works. Let's chat about this over zoom!